### PR TITLE
api server initial commit

### PR DESCRIPTION
**Pull request by slugalisk**

### CODE REVIEW

**Comment by KenanY** [comment]
Can we update to https://github.com/pokowaka/jwt-cpp/commit/48a0f2480706b01f485e4e18095de77db706c9e0? The version in this PR does not support OpenSSL 1.1.0 (admittedly, support was only merged just yesterday).

File: api/deps.sh [hunk 1]
```diff
@@ -29,7 +29,20 @@ apt-get install -y \
...
     libiberty-dev \
-    git-core
+    git-core \
+    wget
+
+echo "installing openssl"
+
+pushd $(/bin/pwd)
+cd /tmp && \
+wget https://www.openssl.org/source/openssl-1.1.0.tar.gz
```
Comment (KenanY) [inline api/deps.sh]: Any particular reason why we shouldn't use a git submodule (to https://github.com/openssl/openssl/commit/577d7010c6ffa98e8dc61d8826693db9d8f57bd5 or whatever) like the rest of the deps?

File: api/src/YoutubeClient.cpp [hunk 2]
```diff
@@ -0,0 +1,159 @@
+#include "YoutubeClient.h"
+
+#include <rapidjson/schema.h>
+#include <rapidjson/stringbuffer.h>
+#include <sstream>
+
```
Comment (KenanY) [inline api/src/YoutubeClient.cpp]: We utilize `std::string` here, so I think we should `#include <string>`

File: api/src/APIHTTPService.h [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+#include "HTTPRequest.h"
+#include "HTTPRouter.h"
+
+namespace rustla2 {
+
+class APIHTTPService {
+ public:
+  explicit APIHTTPService(std::shared_ptr<DB> db);
+
+  void RegisterRoutes(HTTPRouter &router);
```
Comment (KenanY) [inline api/src/APIHTTPService.h]: I think this should this be a `const` reference?

File: api/src/AuthHTTPService.h [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+#include "DB.h"
+#include "HTTPRouter.h"
+#include "TwitchClient.h"
+
+namespace rustla2 {
+
+class AuthHTTPService {
+ public:
+  AuthHTTPService(std::shared_ptr<DB> db);
```
Comment (KenanY) [inline api/src/AuthHTTPService.h]: Can we make this `explicit`?

File: api/src/main.cpp [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+      if (!Listen(hub)) {
+        LOG(FATAL) << "unable to listen";
+      }
+
+      hub.run();
+    });
+  }
+
+  bool Listen(uWS::Hub &hub) {
```
Comment (KenanY) [inline api/src/main.cpp]: Should this be made a `const` reference?

File: api/src/APIHTTPService.h (hunk 3 above)
Comment (paulwe) [reply api/src/APIHTTPService.h]: you're right, this is inconsistent - these should take pointers.

File: api/src/YoutubeClient.cpp (hunk 2 above)
Comment (paulwe) [reply api/src/YoutubeClient.cpp]: it's included int the header

File: api/src/Streams.h [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+    boost::shared_lock<boost::shared_mutex> read_lock(lock_);
+    return update_time_;
+  }
+
+  inline uint64_t GetResetTime() {
+    boost::shared_lock<boost::shared_mutex> read_lock(lock_);
+    return reset_time_;
+  }
+
+  void WriteAPIJSON(rapidjson::Writer<rapidjson::StringBuffer> &writer);
```
Comment (KenanY) [inline api/src/Streams.h]: Make this a `const` reference or a pointer.

File: api/src/Streams.h [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+
+  void WriteJSON(rapidjson::Writer<rapidjson::StringBuffer> &writer);
```
Comment (KenanY) [inline api/src/Streams.h]: Make this a `const` reference or a pointer.

File: api/src/Streams.h [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+  std::string path_;
+  uint64_t viewer_count_{0};
+  uint64_t rustler_count_{0};
+  uint64_t reset_time_{0};
+  uint64_t update_time_{0};
+};
+
+class Streams {
+ public:
+  Streams(sqlite::database db_);
```
Comment (KenanY) [inline api/src/Streams.h]: Make this `explicit`.

File: api/src/Streams.h [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+
+  std::vector<std::shared_ptr<Stream>> GetAllUpdatedSince(uint64_t timestamp);
+
+  std::vector<std::shared_ptr<Stream>> GetAllWithRustlers();
+
+  std::vector<std::shared_ptr<Stream>> GetAllWithRustlersSorted();
+
+  std::string GetAPIJSON();
+
+  void WriteStreamsJSON(rapidjson::Writer<rapidjson::StringBuffer> &writer);
```
Comment (KenanY) [inline api/src/Streams.h]: Make this a `const` reference or a pointer.

File: api/src/WSService.h [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+  ~WSService();
+
+  bool RejectBannedIP(uWS::WebSocket<uWS::SERVER>* ws,
+                      uWS::HttpRequest uws_req);
+
+  void GetStream(uWS::WebSocket<uWS::SERVER>* ws,
+                 const rapidjson::Document& input);
+
+  void GetStreamByID(const uint64_t stream_id,
+                     rapidjson::Writer<rapidjson::StringBuffer>& writer);
```
Comment (KenanY) [inline api/src/WSService.h]: Make this a `const` reference or a pointer.

File: api/src/WSService.h [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+
+  void SetStream(uWS::WebSocket<uWS::SERVER>* ws,
+                 const rapidjson::Document& input);
+
+  inline void SetStreamToChannel(
+      const std::string& channel, const std::string& service,
+      rapidjson::Writer<rapidjson::StringBuffer>& writer, uint64_t* stream_id);
```
Comment (KenanY) [inline api/src/WSService.h]: Make this a `const` reference or a pointer.

File: api/src/WSService.h [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+  void SetStreamToChannel(const std::string& channel,
+                          const std::string& service,
+                          const std::string& overrustle_id,
+                          rapidjson::Writer<rapidjson::StringBuffer>& writer,
```
Comment (KenanY) [inline api/src/WSService.h]: Make this a `const` reference or a pointer.

File: api/src/WSService.h [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+                          uint64_t* stream_id);
+
+  void SetStreamToOverRustleID(
+      const std::string& overrustle_id,
+      rapidjson::Writer<rapidjson::StringBuffer>& writer, uint64_t* stream_id);
```
Comment (KenanY) [inline api/src/WSService.h]: Make this a `const` reference or a pointer.

File: api/src/WSService.h [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+
+  void SetStreamToNull(rapidjson::Writer<rapidjson::StringBuffer>& writer,
```
Comment (KenanY) [inline api/src/WSService.h]: Make this a `const` reference or a pointer.

**Comment by KenanY** [comment]
Everything seems to be fully working locally for me 👍 streams, login, profile updating, banned streams. Fantastic work!

**Comment by slugalisk** [comment]
Thanks! Are we ready to squash this and land it or was there anything else you wanted?

**Comment by KenanY** [comment]
Nah I think it's ready

**Comment by KenanY** [comment]
Landed as https://github.com/ILiedAboutCake/Rustla2/commit/f5226d0b5350c496c9eba9fcb6d4d9db8caee42e

**Pull request closed without merging**